graphs/minimum_spanning_tree_kruskal2.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import annotations
+
+
+class DisjointSetTreeNode:
+    # Disjoint Set Node to store the parent and rank
+    def __init__(self, key: int) -> None:
+        self.key = key
+        self.parent = self
+        self.rank = 0
+
+
+class DisjointSetTree:
+    # Disjoint Set DataStructure
+    def __init__(self):
+        # map from node name to the node object
+        self.map = {}
+
+    def make_set(self, x: int) -> None:
+        # create a new set with x as its member
+        self.map[x] = DisjointSetTreeNode(x)
+
+    def find_set(self, x: int) -> DisjointSetTreeNode:
+        # find the set x belongs to (with path-compression)
+        elem_ref = self.map[x]
+        if elem_ref != elem_ref.parent:
+            elem_ref.parent = self.find_set(elem_ref.parent.key)
+        return elem_ref.parent
+
+    def link(self, x: int, y: int) -> None:
+        # helper function for union operation
+        if x.rank > y.rank:
+            y.parent = x
+        else:
+            x.parent = y
+            if x.rank == y.rank:
+                y.rank += 1
+
+    def union(self, x: int, y: int) -> None:
+        # merge 2 disjoint sets
+        self.link(self.find_set(x), self.find_set(y))
+
+
+class GraphUndirectedWeighted:
+    def __init__(self):
+        # connections: map from the node to the neighbouring nodes (with weights)
+        # nodes: counting the number of nodes in the graph
+        self.connections = {}
+        self.nodes = 0
+
+    def add_node(self, node: int) -> None:
+        # add a node ONLY if its not present in the graph
+        if node not in self.connections:
+            self.connections[node] = {}
+            self.nodes += 1
```

Comment:
Under what circumstances is self.nodes != len(self.connections)?

Reply:
Yeah its always same. Should I remove it?

Reply:
Your call.  You already have my approval so you can squash & merge whenever you want (as long as the tests are green).  Well done!

Reply:
I have made the changes, but I cannot merge since I'm not a member